Replace `DataLocale` by `Locale` on API, `LanguageIdentifier` on `DataRequest`

robertbastian opened the pull request:
Unicode extension keys are now modeled as `DataKeyAttributes`.

Collator is currently failing because I haven't looked at fallback yet.

sffc commented on components/calendar/src/week_of.rs:
```diff
@@ -52,20 +52,20 @@ impl WeekCalculator {
     /// [📚 Help choosing a constructor](icu_provider::constructors)
     #[cfg(feature = "compiled_data")]
-    pub fn try_new(locale: &DataLocale) -> Result<Self, CalendarError> {
+    pub fn try_new(locale: &Locale) -> Result<Self, CalendarError> {
```
Issue: This is a regressive change.

1. We have evidence that `Locale` is bigger and slower than `DataLocale`
2. We are migrating to `Preferences`, and the `.into()` will make that change more likely to "just work" when it is ready to land.

(I thought we were aligned on this. I'm surprised to see this PR in my inbox.)

robertbastian replied:
This is a step in this direction. The next step will be to replace `Locale` by `Preferences`, but at some point we have to decouple the API type from the provider type.

robertbastian replied:
> 1. We have evidence that Locale is bigger and slower than DataLocale

This only uses `Locale` on the API boundary, I then directly borrow the `LanguageIdentifier` and `DataKeyAttributes`, which should be both smaller and faster than the current model.

robertbastian replied:
I changed to title to emphasize the important change, getting `DataRequest` to the design in https://github.com/unicode-org/icu4x/issues/3632

robertbastian replied:
So it sounds like you want to keep `DataLocale` on the API so that we don't remove the `into()`s now? I can do that

sffc replied:
I think my preferred transitional state would be keeping DataLocale and all public call sites the same, adding a function to DataLocale that returns &LanguageIdentifier, and using it when constructing the DataRequest

robertbastian replied:
This is basically what I've done, but I've made DataLocale an open struct and deleted all of its logic.

sffc commented on components/calendar/src/any_calendar.rs:
```diff
@@ -1073,33 +1073,24 @@ impl AnyCalendarKind {
+            .unwrap_or_else(|| {
+                if l.id.language == language!("th") {
+                    Self::Buddhist
+                // Other known fallback routes for currently-unsupported calendars
+                // } else if lang == language!("sa") {
+                //     Self::IslamicUmalqura
+                // } else if lang == language!("af") || lang == language!("ir") {
+                //     Self::Persian
```
Observation: These calendars have been supported since 1.4 but this seems to have gotten lost at some point; we should make sure there's an issue tracking it

sffc reviewed (commented):
Praise: It looks approximately what I expect to see but it's a big PR so I haven't finished looking at every file

sffc commented on components/collator/src/comparison.rs:
```diff
@@ -145,25 +146,61 @@ impl Collator {
 
-        let metadata_payload: DataPayload<crate::provider::CollationMetadataV1Marker> =
-            provider.load(req)?.take_payload()?;
+        let metadata_payload: DataPayload<crate::provider::CollationMetadataV1Marker> = provider
+            .load(req)
+            .or_else(|_| {
+                provider.load(DataRequest {
+                    langid: req.langid,
+                    ..Default::default()
+                })
+            })?
+            .take_payload()?;
```
Suggestion, here and elsewhere: perhaps refactor to make more clear why you are making two requests

```suggestion
        let metadata_payload: DataPayload<crate::provider::CollationMetadataV1Marker> = provider
            .load(req_specific)
            .or_else(|_| {
                provider.load(req_default)
            })?
            .take_payload()?;
```

sffc commented on components/datetime/tests/datetime.rs:
```diff
@@ -486,15 +474,15 @@ fn test_dayperiod_patterns() {
                             *expected,
                             "\n\
-                            locale:   `{}`,\n\
+                            locale:   `{:?}`,\n\
```
Thought/Issue: We probably want to keep a `Display` impl on Preferences?

robertbastian replied:
I don't want to make that assumption now. It's not a problem to use `Debug` in a test.

sffc commented on components/locale/src/fallback/algorithms.rs:
```diff
@@ -56,36 +51,23 @@ impl<'a> LocaleFallbackerWithConfig<'a> {
-        // 3. Remove irrelevant extension subtags
-        locale.retain_unicode_ext(|key| {
-            match *key {
-                // Always retain -u-sd
-                SUBDIVISION_KEY => true,
```
Thought/Issue: We don't currently use `-u-sd` but it would be nice to support it in LocaleFallbacker for future compatibility

robertbastian replied:
This is a new requirement, please don't block the PR on this.

sffc replied:
I won't block the PR on it but it's not a new requirement; the old fallbacker supported -u-sd so it's a regression in behavior that it doesn't support it now and doesn't have a clear path to supporting it...

robertbastian replied:
We had a design discussion in March to drop unicode keywords from the fallbacker and this was not brought up since then. This will require a lot of rethinking.

sffc commented on components/locale/src/fallback/algorithms.rs:
```diff
@@ -229,101 +175,23 @@ mod tests {
-            expected_language_chain: &["en-US-u-sd-usca", "en-US", "en-u-sd-usca", "en"],
-            expected_region_chain: &["en-US-u-sd-usca", "en-US", "und-US-u-sd-usca", "und-US"],
-        },
-        TestCase {
-            // TODO(#4413): -u-rg is not yet supported; when it is, this test should be updated
-            input: "en-u-rg-gbxxxx",
-            requires_data: false,
-            extension_key: None,
-            fallback_supplement: None,
-            expected_language_chain: &["en"],
-            expected_region_chain: &["en"],
-        },
```
Issue: We should fix this in 2.0 because it might require an architecture change in LocaleFallabcker. https://github.com/unicode-org/icu4x/issues/4413

sffc commented on provider/core/src/hello_world.rs:
```diff
@@ -111,12 +111,12 @@ impl HelloWorldProvider {
+        ("en-019", "", "Hello from 🌎"),       // AMERICAS
+        ("en-142", "", "Hello from 🌏"),       // ASIA
+        ("en-GB", "", "Hello from 🇬🇧"),        // GREAT BRITAIN
+        ("en-GB", "england", "Hello from 🏴󠁧󠁢󠁥󠁮󠁧󠁿"), // ENGLAND
```
Issue: `england` should remain a subdivision, not a key attribute... please at least add a TODO(#4413)

robertbastian replied:
This is as designed

sffc reviewed (commented):
Down through `provider/core`; next stop `provider/datagen`

sffc commented:
Reasons we may want to keep `DataLocale` and not use `LanguageIdentifier`:

1. We can make it `Copy` by supporting only a single variant
2. We can support subdivision
3. We can add other helper functions and are not bound to the constraints of `LanguageIdentifier`

robertbastian commented:
> Reasons we may want to keep DataLocale and not use LanguageIdentifier:

We had agreed on this and I have implemented this, which was a non-trivial amount of work. 

> We can make it Copy by supporting only a single variant

What does this gain us? 

> We can support subdivision

Ok, but there are other ways as well

> We can add other helper functions and are not bound to the constraints of LanguageIdentifier

Having logic duplicated between `DataLocale` and `icu_locale` really bothered me about the old `DataLocale`, because that was code duplication, as well as double the API surface, with subtle API differences. I don't see `LanguageIdentifier` as a constraint, if we need more functionality we should add it there so that other users can benefit from it (like I added `is_und` in this PR).

sffc commented:
^ If we can demonstrate that there's a good way to support `-u-sd` and `-u-rg` in this design, I'm onboard

Manishearth commented:
(currently feel like no strong opinion on whether we should land this. happy to take part in discussions, but no need to wait for my review, quickly going through this I don't see a reason why I would _block_ this)

sffc commented:
Current thinking of @sffc @zbraniecki:

- Subtags could be about preferences/hints and data identity.
- `-u-sd` is either a preference (I want Scottish unit preferences) or a data identity (I want Catalan as spoken in Valencia)
- Variants are a grab bag that encode different concepts such as orthography (POLYTON), subdivision (VALENCIA), and operating system (MACOS), and probably other things, too. ([full list](https://github.com/unicode-org/cldr/blob/maint/maint-41/common/validity/variant.xml))
- @zbraniecki notes that subdivision suffixes are 1-3 alphanums. This means in principle we could encode them in an ICU4X language identifier following a region, such as `en-US-ca`, without risking conflict with variants or other subtags. When converting to Locale, we make it `en-US-u-sd-usca`, and when converting to LanguageIdentifier, we drop the subdivision. `ca_ES_VALENCIA` would eventually become `ca-ES-vc`. Would also be nice to drop the variant subtag.
- We could consider using this new type in Preferences. Multi-variants would be mapped to their respective proper keywords, with keywords added as necessary.
- @sffc notes that variants could be used by prepending "sd", such as `en-US-sdusca` (guaranteed to be 5-7 alphanums), if we wanted to stick with the traditional LanguageIdentifier representation.
- Would like to bring this to the CLDR Design WG for further discussion.

sffc commented:
Design doc: https://docs.google.com/document/d/1Rrp1IxycTu2CETceUOQRcwVECXsv0pOJIAUMPdYlHYY/edit

zbraniecki commented:
Subtle adjustment:

> -u-sd is either a preference (I want Scottish unit preferences) or a data identity (This is Valencian data)

sffc commented:
We discussed this at the CLDR design meeting. There wasn't a clear conclusion on what fields should be in data locale, but it was evident that this could change over time. An approach that seems like it would make sense for us would be to restrict to a single subtag on top of LSR which contains an implementation-defined enumeration of data identity. For example, we lump all variants and subdivisions into our own namespace.

robertbastian commented:
Discussion

```rust
#[derive(Copy, ...)]
pub struct DataLocale {
    language: icu_locale_core::subtags::Language,
    script: Option<icu_locale_core::subtags::Script>,
    region: Option<icu_locale_core::subtags::Region>,
    subdivision: Option<icu_locale_core::extensions::unicode::SubdivisionSuffix>,
    single_variant: Option<icu_locale_core::subtags::Variant>,
}
```
* simplify `DataLocale` API by removing extensions stuff
* have getters and setters for the singleton fields, as well as `is_und`/`is_empty`/`is_default`
* move to `icu_preferences`?
* fallback still uses `DataLocale`
